### ISSUE TITLE
Questionable Score Fitting

### ISSUE THREAD

**Issue by jeff-hykin** [EpistasisLab/tpot#1320]
TLDR; a scoring method is chosen, but it does not seem to be followed.

## Context of the issue

I have a skewed dataset. To help with that, I'm using a very simple scoring function (which I've confirmed is getting called/run) but tpot does not seem to follow it. Example behavior: when given 60% accuracy on true-positives of `class_1` and  99% accuracy on true-positives of `class_2` the scoring function returns `0.6` (the minimum of the two)

<img width="834" alt="Screen Shot 2023-09-12 at 11 22 56 AM" src="https://github.com/EpistasisLab/tpot/assets/17692058/92c11f29-64d9-46e2-b2b9-1a2dfeec4f54">

Here's some outputs of that^ when it is run with tpot:

<img width="645" alt="Screen Shot 2023-09-12 at 11 25 15 AM" src="https://github.com/EpistasisLab/tpot/assets/17692058/f1a6ee80-7aa8-4c18-b4dc-29c630d34d53">

Tpot is being called like this
- [the config](https://github.com/ioerger2/DNG1/blob/1d296ef2a41ff05d70a8c997eeb4c7a06d31fac9/main/model_tpot.py#L30) is basic, mostly from templates.
- [tpot log/output](https://github.com/ioerger2/DNG1/blob/76dd01dc4e4e4e29baa1d1b81eb02e5476fa3feb/main/results/recent_tpot.log#L3) (23 hour run, should be comparable to the 2 day run mentioned below)

<img width="437" alt="Screen Shot 2023-09-11 at 8 10 39 AM" src="https://github.com/EpistasisLab/tpot/assets/17692058/5fe59d9f-c8f8-4772-91c4-6f8051b78af9">

## Symptom 1

Tpot seems to be strategically picking worse values (even though scorer has `greater_is_better=True`)

- training for 15 min gets about 14% positive accuracy (using balanced_accuracy repeated 50 times to get an average)
- training for 2 days gets about 12% positive accuracy (using balanced_accuracy OR custom equal_accuracy_scorer)

Training for 15 minutes:
<img width="527" alt="Screen Shot 2023-09-11 at 8 20 57 AM" src="https://github.com/EpistasisLab/tpot/assets/17692058/b4de620e-cb29-44d6-9c51-f93380d6dc70">

Training for 2 days:
<img width="769" alt="Screen Shot 2023-09-11 at 8 25 23 AM" src="https://github.com/EpistasisLab/tpot/assets/17692058/e8db3bd1-e077-443e-a064-24c279fd48a9">

## Symptom 2

Its easy to get 60% positive accuracy, but tpot never does.

Random forests are part of my tpot config. However, I can manually use a random forest, along with ANY combination of parameters from the tpot config, and it blows the 2-day tpot training out of the water (72% positive accuracy instead of 12%, and also better overall accuracy)

E.g. here's the tpot config for random forests
<img width="518" alt="Screen Shot 2023-09-11 at 8 31 31 AM" src="https://github.com/EpistasisLab/tpot/assets/17692058/c28a24c8-3598-4789-848c-7ee7848090e0">

Here's how I replace tpot with a random forest directly
<img width="886" alt="Screen Shot 2023-09-11 at 8 31 22 AM" src="https://github.com/EpistasisLab/tpot/assets/17692058/ffe38317-0986-47ba-bdb1-126c4e59384e">

Here's the resulting performance:
<img width="783" alt="Screen Shot 2023-09-11 at 8 35 23 AM" src="https://github.com/EpistasisLab/tpot/assets/17692058/1542c470-4651-44d3-9ff4-10cb4ae77dcb">

## Conclusion

Am I doing something wrong?

The scoring function is called but it doesn't seemed to be used.

**Comment by perib** [comment]
Your links go to a 404 error page; the repository is probably private. Your screenshot shows that one of the CV scores has an expected value of around .68. (It would be helpful to copy/paste your code rather than use screenshots, so we can quickly test it ourselves). 

Is the lower accuracy you refer to on the out-of-sample test set? Sometimes TPOT can overfit the CV score with overly complex pipelines. So, while it is "correctly" optimizing the objective function, the result is a pipeline that performs poorly on held-out data. One option is that you could look at the pareto_front_fitted_pipelines_ to see if any of those simpler pipelines have better performance. (For example, You could hold out a validation set, use that to select from the Pareto front, and then do a final test on that dataset.) By default tpot uses 5 fold CV, you could try setting cv=10. You can also minimize complexity with template="selector-transformer-classifier"

When I run the TPOT with a test dataset with this scorer, it seems to accurately maximize the objective function.

**Comment by jeff-hykin** [comment]
Thanks for the reply, and sorry about the links, I forgot the development repo was private.

I think you're right, and yes there is a held-out dataset for the final evaluation score. I was thinking 5 fold CV within tpot would be enough to detect/prevent overfitting, but the training set does have very few positive examples so I probably should've been looking for it. I'll run some tests and if that is the case, then this is definitely non-issue (and sorry about that). If it is the case, maybe I'll add some overfitting detection or limited-examples-warning in a PR.

As for reproducing, no worries, I was just looking for high-level feedback first (rather than a full reproduction/debugging/analysis). If does appear to be a bug, I'll put in the work to isolate the dataset+code and make it fully reproducable on a public repo.

**Comment by perib** [comment]
We would appreciate the PR! We have moved development to focus on TPOT2, which you can find here: https://github.com/EpistasisLab/tpot2 . It was rebuilt from the ground up to be more modular and easier to develop. 

In TPOT2, we have the parameters validation_fraction and validation_strategy to try to address this issue. It can hold out x% of the training data, which it will use as a validation set for the pareto front models.

**Comment by jeff-hykin** [comment]
Alright, I confirmed sample size was the issue (e.g. no issue for tpot).

Not only that but tpot worked amazing! I was able to get +10% on an f1 score over the best hand-crafted architecture for this problem.

Thanks again for the help

**Comment by perib** [comment]
great! Glad the issue has been resolved. I would also recommend trying out our next version of TPOT, TPOT2 found here: https://github.com/EpistasisLab/tpot2
